Improvements to bumpversion, circleci, and packagecloud

jeremycarroll opened the pull request:
When I used the first version (syapse-ontology) from this cookiecutter that were some minor issues.
Most significant is bumpversion changes, which is improved (see the [README](https://github.com/syapse/cookiecutter-pypackage/blob/f10ba417343532e533adba3cad80657dfe5969ab/%7B%7Bcookiecutter.project_dash_slug%7D%7D/README.md#versioning)).
Circleci bug fix and packagecloud improvement too.


@ganewman42 @eabruzzese The change to the post gen hook correctly computes a read token from package cloud, so we have a different read token for each package generated from this cookie cutter.

Commit:
Improvements to bumpversion, circleci, and packagecloud from ontology package

Commit:
Punctuation

kojiromike commented on README.md:
```diff
@@ -26,11 +26,12 @@ $ docker build -t foo . && docker run -it foo
+| `project_short_description`   | A few words about the project.                                                           |
+| `version`                     | Semantic Version number.                                                                 |
+| `command_line_interface`      | Whether to provide a command line tool.                                                  |
+| `optional_packagecloud_token` | The token for accessing package cloud. (Do not include in this __public__ repo.)         |
```
Because this is a security credential I'm opposed to exposing this on the commandline. I'd prefer it to be used exclusively via the environment (basically, for CircleCI contexts), but if it's needed I wouldn't be opposed to prompting for it in a non-echoing prompt.

Commit:
AP-26686: package cloud read token

Commit:
AP-26686: documentation update

eabruzzese commented:
@jeremycarroll This looks pretty good, and I have a few suggestions for this that I'll bring up in our meeting today if that's okay. I'll try to prototype my thoughts before then as well.

kojiromike commented on {{cookiecutter.project_dash_slug}}/.circleci/config.yml:
```diff
@@ -47,8 +47,11 @@ default: &steps
+              ls -1 TEST-*.xml | awk -F. -f move-results.awk | bash
+            }
+            pipenv run python -m xmlrunner discover -s tests -t . -p '*.py' || (save && false)
+            save
```
This is pretty opaque. Please make it simpler.

eabruzzese replied:
@jeremycarroll can you help me understand what this code is supposed to do? I don't know awk very well.

jeremycarroll replied:
I think I should recode in python

eabruzzese committed:
Refactor packagecloud integration

The cookiecutter will always ask for a packagecloud repo and master
token. This is a safe assumption (since the cookiecutter is only used
within Syapse) and it's low-risk even if they don't pull anything from
packagecloud. Having the additional source available will lower the
barrier to entry for using other Syapse packages later.

The logic for configuring packagecloud authentication was moved into a
`bin/setup` script. This ensures that we don't break `cookiecutter`'s
`--no-input` and `--replay` options.

The `bin/setup` command essentially writes to a `.env` file which is
sourced by `pipenv` when a developer wants to work on the project. The
script will prompt them for a token if `PACKAGECLOUD_TOKEN` isn't set.

eabruzzese committed:
Include an option for removing packagecloud.

The `enable_packagecloud` flag was added to `cookiecutter.json`. If
disabled, it won't generate any packagecloud-related code.

eabruzzese committed:
Skip packagecloud config if the read token is set

eabruzzese committed:
Merge pull request #4 from syapse/packagecloud-configuration

Refactor packagecloud integration

Commit:
Python 2.7 optional AP-26814

Commit:
Excised tox support from generated project AP-26815

Commit:
Merge pull request #5 from syapse/remove-tox-optional-2.7

Remove tox and make python 2.7 optional

jeremycarroll commented:
Note despite @mianelson 's approval we have the following unresolved issues:
1.  opaqueness of the testing (particularly the awk)
1.  on-going discussion between myself and @eabruzzese concerning the details of setup and the token

Commit:
Merge branch 'develop' into update-from-first-package

# Conflicts:
#	README.md

Commit:
Copied from https://github.com/syapse/syapse-ontology/pull/8

Commit:
Reworking the token to prompt on creation and store in repo. Support --no-input and --replay

jeremycarroll commented:
@eabruzzese I rearranged the code from your PR so that the generation of read token happens at cookie cutting time; and is stored in the new repo (like with my original code) (and also in the replay template)
I addressed your comments about `--no-input` and `--replay` and have tested both options.
I kept the `setup` from your PR, but not the read token part. Hence out of the box, running it is optional.

Commit:
Instructions from Eric.
